src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/PostcodeSearch.cs

```diff
@@ -0,0 +1,194 @@
+            var validationResult = validator.Validate(request);
+
+            if (!validationResult.IsValid)
+            {
+                return new ModelWithErrors<SearchCommand>(request, validationResult);
+            }
+
+            // We need the postcode to exist in ONSPD so we can resolve a lat/lng.
+            // Check it's present before we do an address search so we can output an error early.
+
+            var onspdSearchResult = await _onspdSearchClient.Search(new OnspdSearchQuery() { Postcode = request.Postcode });
+
+            if (onspdSearchResult.Items.Count == 0)
+            {
+                return CreateInvalidPostcodeResponse();
+            }
+
+            var vm = await CreateViewModel(request.Postcode);
+
+            if (vm.Results.Count == 0)
+            {
+                return CreateInvalidPostcodeResponse();
+            }
+
+            return vm;
+
+            ModelWithErrors<SearchCommand> CreateInvalidPostcodeResponse() =>
+                new ModelWithErrors<SearchCommand>(
+                    request,
+                    new ValidationResult(new[]
+                    {
+                        new ValidationFailure(nameof(SearchCommand.Postcode), "Enter a real postcode")
+                    }));
+        }
+
+        public async Task<OneOf<ModelWithErrors<SearchViewModel>, Success>> Handle(
+            SelectCommand request,
+            CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            var validator = new SelectCommandValidator();
+            var validationResult = validator.Validate(request);
+
+            if (!validationResult.IsValid)
+            {
+                // No address specified. Re-run the search with a prompt to select an address.
+                var vm = await CreateViewModel(request.Postcode);
+                return new ModelWithErrors<SearchViewModel>(vm, validationResult);
+            }
+
+            var addressDetail = await _addressSearchService.GetById(request.AddressId);
+
+            if (addressDetail == null)
+            {
+                // Address specified isn't valid. (We should never get here.)
+                // Re-run the search to get a list of valid addresses with a prompt to select an address.
+
+                var badAddressValidationResult = new ValidationResult(new[]
+                {
+                    new ValidationFailure(nameof(SelectCommand.AddressId), "Select an address")
+                });
+
+                var vm = await CreateViewModel(request.Postcode);
+                return new ModelWithErrors<SearchViewModel>(vm, badAddressValidationResult);
+            }
+
+            // SearchCommand has already checked that there is an ONSPD entry for this postcode
+            // so there is no valid case where `.Single()` should blow up
+            var onspdSearchResult = await _onspdSearchClient.Search(new OnspdSearchQuery() { Postcode = addressDetail.Postcode });
```

Comment:
Why are we running it again? Surely this data could've been added to the journeystate?

Reply:
Done.